[Solutions] Upgrade/downgrade the deployed solutions from UI

ChengYanJin opened the pull request:
**Component**: ui, solutions

**Context**:  Currently, in order to upgrade the deployed solutions the only way is to click on the add button and fill in the form with the new version, which is counter-intuitive.

**Summary**:
- We can access the solution detail page from the environment list by click on the specific row
- Upgrade & Downgrade button should appear only if there are some available versions to upgrade / downgrade.
- When click on the `upgrade/downgrade` button, it will pop up a modal with available upgrade/downgrade version 
- Rename the sidebar from `solutions` to `environment` as well as the URL from `/solutions` to `/evironments` 

**Acceptance criteria**: 

1. Environment list page

 ![image](https://user-images.githubusercontent.com/18453133/78365299-3238ad80-75bf-11ea-9f60-1ed5ad8d4944.png)
By clicking on the specific environment `dev` for example,  we will go to the dev detail page, given the condition that we already deployed at least one solution in this environment.

2. Solution detail page

![image](https://user-images.githubusercontent.com/18453133/78348274-09ee8600-75a2-11ea-9e53-6182195458ce.png)

3. Upgrade/downgrade a solution

![image](https://user-images.githubusercontent.com/18453133/78365325-3f559c80-75bf-11ea-8ece-540a1c83fe14.png)

After clicking on the`Upgrade/Downgrade` button, there will be a modal pop up to edit the deployed solution version

---

<!-- Declare one or more issues to close once this PR gets merged -->

Closes: #2337 

<!-- If you want to refer to an issue while not closing it, use:

See: #ISSUE_NUMBER

-->

ChengYanJin committed:
ui/solutions: Extract the the table based page container style into a separate file

Since we have the same pageContainer style duplicated in several components, it makes sense
to extract the style into a separate file.

Refs: #2337

ChengYanJin committed:
ui/solutions: Add the solutionDetail page

When we click on one environment, if there is solution deployed in this env, then we will redirect to the solutionDetail page.

Refs: #2337

ChengYanJin committed:
ui/solutions: Rename the sidebar, url, breadcrumb from solutions to environment

Since we create an environment detail page, the url make more sense to /environment/`${env_name}`

Refs: #2337

ChengYanJin committed:
ui/solutions: Update the solutions list with availableUpgradeVersion and availableDowngradeVersion

when updateEnvironments update the solution list

Refs: #2337

ChengYanJin committed:
ui/solutions: Update the solutionList Container

Add the check before going to the solutionDetail page

Refs: #2337

ChengYanJin committed:
ui/solutions: Update the translations

Refs: #2337

ChengYanJin committed:
ui/solutions: Extract the common style for modal form

Refs: #2337

bert-e commented:
# Hello chengyanjin,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort

thomasdanan reviewed (commented):
how and when do you choose to which version you want to upgrade or downgrade?

ChengYanJin commented:
> how and when do you choose to which version you want to upgrade or downgrade?

When you click on the upgrade/downgrade button, it will pop up a modal with the available version.
(I'll update a screenshot for it)

thomasdanan commented:
also don't hesitate to add @xaf-scality @bkettler to review the PR

ChengYanJin committed:
ui/test: Update the test for solution saga

Refs: #2337

ChengYanJin committed:
ui/solutions: Check the deploy version of UI&Operator before update the environment solutions

- We may encounter a failure when deploy a new solution or upgrade/downgrade the solution,
  so when we update the Environment  we should check the deployment version instead of retrieve it from configmap
- In order to check the deployment version of Operato&UI, the only way we have currently is check the image verion,
  but we should add a label with the solution version in both Operator & UI (will do it in a debt ticket)

Refs: #2337

gdemonet commented:
I'm a bit worried with the "Upgrade" / "Downgrade" buttons:

- there's no real difference between the two, so not sure there's much value in having two buttons, a single one with "Edit" label could be sufficient (could also allow for more editions than just the version, in the future)
- it's not going to upgrade or downgrade the components deployed, only the Operator and UI - making this information obvious in our UI would make sense, to avoid unnecessary confusions (not sure how though)

gdemonet commented on ui/src/ducks/app/solutions.js:
```diff
@@ -204,16 +204,41 @@ export function* updateEnvironments(environments) {
         );
         if (!solutionOperatorDeployment.error && !solutionUIDeployment.error) {
+          const currentSolutionVersion = envConfig?.[solution];
+          // when update the environment, we add `availableUpgradeVersion` and `availableDowngradeVersion` field
```
We shouldn't do this kind of thing, it's duplicating existing data we have in the state. We should just make sure whenever it's consumed that we know how to compute these lists from a selected version and all the available ones (that we have stored in the state).

So, the code you wrote here should either go directly in the presentation component, or as a utility method. But the computation isn't expensive, so there's no need to cache the result in Redux.

ChengYanJin replied:
Hello @gdemonet,

So what you mean is to store `availableUpgradeVersion` and `availableDowngradeVersion` data in the local component state instead of caching it in redux store?

gdemonet replied:
Yeah, that's the idea. A single row can compute whatever it needs from its props and the list of available solutions in the global state :)

gdemonet commented on ui/src/ducks/app/solutions.js:
```diff
@@ -203,9 +203,22 @@ export function* updateEnvironments(environments) {
+
         if (!solutionOperatorDeployment.error && !solutionUIDeployment.error) {
-          const currentSolutionVersion = envConfig?.[solution];
+          // in order to get the deployed version, we temporarily check the image version deployed Operator & UI.
+          // but we should add a label with the solution version in both Operator & UI.
```
Labels were added in #2367 ;)

gdemonet commented on ui/src/ducks/app/solutions.js:
```diff
@@ -217,26 +230,42 @@ export function* updateEnvironments(environments) {
-            if (avaSolVer.version.localeCompare(currentSolutionVersion) <= -1) {
+            if (
+              avaSolVer.version.localeCompare(
+                isDeploySolutionSucceed
+                  ? currentSolutionVersion
+                  : solutionOperatorDeploymentVersion,
```
No need for this check, both branches always yield the `solutionOperatorDeploymentVersion` anyway.

What would like to show differently if the configured version is different from the deployed one? For me, the only  thing to show would be that I can only see a "apply" button if there's a difference between `current` and `observed` versions.

ChengYanJin replied:
Yes, you are right! We should go for Operator Version anyways because it's the actual deployed version.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort





The following reviewers are expecting changes from the author, or must review again:

* @gdemonet

ChengYanJin committed:
ui/solution: Remove the availableUpgradeVersion and availableDowngradeVersion from the global store

We shouldn't store duplicate data in the global redux store.

Refs: #2337

ChengYanJin committed:
ui/solutions: Get the deployed solution version from the labels

Refs: #2337

ChengYanJin committed:
ui/cypresstest: Check the operator deployment instead of ui deployment

Since UI deployment is optional

Refs: #2337

gdemonet commented on ui/src/services/k8s/solutions.js:
```diff
@@ -3,7 +3,7 @@ import * as core from './core';
 const _K8S = 'app.kubernetes.io';
 const _METAL = 'solutions.metalk8s.scality.com';
-
+const LABEL_VERSION = 'app.kubernetes.io/version';
```
super minor:

```suggestion
const LABEL_K8S_VERSION = `${_K8S}/version`;
```

gdemonet reviewed (dismissed):
LGTM!

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @nootal

* @alexandre-allard-scality

* @lucieleonard

* @sayf-eddine-scality

* @slaperche-scality

* @nicolast

* @gdemonet

* @carlito-scality

* @thomasdanan

* @jbertran

* @teddyandrieux

* @cmonfort

* @monpote

ChengYanJin committed:
ui/solutions: Check the existence deployed version before display in the SolutionList

bugfix

Refs: #2337

ChengYanJin commented:
/approve

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.6`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.6`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`



Please check the status of the associated issue None.

Goodbye chengyanjin.
